File: backend/src/middleware/rbac.ts
import type { NextFunction, Request, Response } from "express";
import type { UserType } from "@prisma/client";
import { ApiError } from "../utils/ApiError";

/** Coarse-grained gate by account type (customer / rental partner / admin / super admin). */
export function requireUserType(...allowed: UserType[]) {
  return (req: Request, _res: Response, next: NextFunction) => {
    if (!req.user) return next(ApiError.unauthorized());
    if (!allowed.includes(req.user.userType)) {
      return next(ApiError.forbidden("Your account type cannot access this resource"));
    }
    return next();
  };
}

/** Fine-grained gate for admin/staff roles carrying explicit permission keys (e.g. "vehicles.approve"). */
export function requirePermission(...permissionKeys: string[]) {
  return (req: Request, _res: Response, next: NextFunction) => {
    if (!req.user) return next(ApiError.unauthorized());
    if (req.user.userType === "SUPER_ADMIN") return next();

    const hasAll = permissionKeys.every((key) => req.user!.permissions.includes(key));
    if (!hasAll) {
      return next(ApiError.forbidden("You do not have permission to perform this action"));
    }
    return next();
  };
}

File: backend/app/deps/rbac.py
from collections.abc import Callable

from fastapi import Depends

from app.core.responses import ApiError
from app.db.enums import UserType
from app.deps.auth import AuthUser, get_current_user


def require_user_type(*allowed: UserType) -> Callable[..., AuthUser]:
    """Coarse-grained gate by account type (customer / rental partner / admin / super admin)."""

    async def dependency(user: AuthUser = Depends(get_current_user)) -> AuthUser:
        if user.user_type not in allowed:
            raise ApiError.forbidden("Your account type cannot access this resource")
        return user

    return dependency


def require_permission(*permission_keys: str) -> Callable[..., AuthUser]:
    """Fine-grained gate for admin/staff roles carrying explicit permission keys."""

    async def dependency(user: AuthUser = Depends(get_current_user)) -> AuthUser:
        if user.user_type == UserType.SUPER_ADMIN:
            return user
        if not all(key in user.permissions for key in permission_keys):
            raise ApiError.forbidden("You do not have permission to perform this action")
        return user

    return dependency
